internal/graph/cached_resolver_test.go

```diff
@@ -515,3 +513,214 @@ func TestCachedCheckDatastoreQueryCount(t *testing.T) {
+}
+
+func TestCheckCacheKeyWithContextualTuples(t *testing.T) {
+	storeID := ulid.Make().String()
+	modelID := ulid.Make().String()
+
+	tuples1 := []*openfgav1.TupleKey{
+		tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+		tuple.NewTupleKey("document:2", "viewer", "user:jon"),
+	}
+
+	tuples2 := []*openfgav1.TupleKey{
+		tuple.NewTupleKey("document:2", "viewer", "user:jon"),
+		tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+	}
+
+	key1, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:x", "viewer", "user:jon"),
+		ContextualTuples:     tuples1,
+	})
+	require.NoError(t, err)
+
+	key2, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:x", "viewer", "user:jon"),
+		ContextualTuples:     tuples2,
+	})
+	require.NoError(t, err)
+
+	require.Equal(t, key1, key2)
+}
+
+func TestCheckCacheKeyWithContext(t *testing.T) {
+	storeID := ulid.Make().String()
+	modelID := ulid.Make().String()
+
+	struct1, err := structpb.NewStruct(map[string]interface{}{
+		"key1": "foo",
+		"key2": "bar",
+	})
+	require.NoError(t, err)
+
+	struct2, err := structpb.NewStruct(map[string]interface{}{
+		"key2": "bar",
+		"key1": "foo",
+	})
+	require.NoError(t, err)
+
+	struct3, err := structpb.NewStruct(map[string]interface{}{
+		"key2": "x",
+		"key1": "foo",
+	})
+	require.NoError(t, err)
+
+	struct4, err := structpb.NewStruct(map[string]interface{}{
+		"key2": "x",
+		"key1": true,
+	})
+	require.NoError(t, err)
+
+	key1, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+		Context:              struct1,
+	})
+	require.NoError(t, err)
+
+	key2, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+		Context:              struct2,
+	})
+	require.NoError(t, err)
+
+	key3, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+		Context:              struct3,
+	})
+	require.NoError(t, err)
+
+	key4, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+		Context:              struct4,
+	})
+	require.NoError(t, err)
+
+	require.Equal(t, key1, key2)
+	require.NotEqual(t, key1, key3)
+	require.NotEqual(t, key1, key4)
+	require.NotEqual(t, key3, key4)
+}
+
+var checkCacheKey string
+
+func BenchmarkCheckRequestCacheKey(b *testing.B) {
```

Comment:
@jon-whit it'd be nice to lift this benchmark up to our main branch just for comparison sake.

Reply:
https://github.com/openfga/openfga/pull/1095